File: bindings/imgui_bundle/demos_cpp/demo_utils/api_demos.cpp
#include "api_demos.h"
#include "imgui_md_wrapper.h"
#include "immapp/immapp.h"
#include "immapp/code_utils.h"
#include "immapp/snippets.h"
#include "imgui_md_wrapper/imgui_md_wrapper.h"
#include "demo_utils/subprocess.h"
#include "hello_imgui/internal/whereami/whereami_cpp.h"

#include <fplus/fplus.hpp>
#include <unordered_map>
#include <string>
#include <functional>
#include <filesystem>


std::string DemosAssetsFolder()
{
#ifndef __EMSCRIPTEN__
    return "demos_assets/";
#else
    return "/";
#endif
}


void ChdirBesideAssetsFolder()
{
    auto findDemoAssets = [](const std::filesystem::path& path) -> bool
    {
        std::filesystem::path candidateAssetsFolder = path / DemosAssetsFolder();
        if (std::filesystem::is_directory(candidateAssetsFolder))
        {
            // chdir to the assets folder parent
            std::filesystem::current_path(path);
            if (! std::filesystem::is_directory(DemosAssetsFolder()))
                throw std::runtime_error("ChdirBesideAssetsFolder => fail setting path");
            HelloImGui::SetAssetsFolder(DemosAssetsFolder());
            return true;
        }
        else
            return false;
    };

    // 1. Try to find demo assets in current folder
    const auto& currentPath = std::filesystem::current_path();
    if ( findDemoAssets(currentPath) )
        return;

    // 2. Try to find demo assets in current folder parent
    if (findDemoAssets(currentPath.parent_path()))
        return;


    #ifdef HELLOIMGUI_INSIDE_APPLE_BUNDLE
    {
        std::filesystem::path exeFolder(wai_getExecutableFolder_string());
        auto apps_bundles_path = exeFolder.parent_path().parent_path().parent_path();

        if (findDemoAssets(apps_bundles_path))
            return;
    };
    #endif

    #ifndef __EMSCRIPTEN__
    // 3. Try to find demo assets in exe folder
    {
        std::filesystem::path exeFolder(wai_getExecutableFolder_string());
        if (findDemoAssets(exeFolder))
            return;
        // 4. Try to find demo assets in exe folder parent (for MSVC Debug/ and Release/ folders)
        if (findDemoAssets(exeFolder.parent_path()))
            return;
    }
    #endif

    std::cerr << "Could not find " << DemosAssetsFolder() << " folder!\n";
}



std::string MainPythonPackageFolder()
{
#ifdef __EMSCRIPTEN__
    return "/";
#else
    auto thisDir = std::filesystem::path(__FILE__).parent_path();
    auto grandParentDir = thisDir.parent_path().parent_path();
    return grandParentDir.string();
#endif
}

std::string DemoCppFolder()  { return MainPythonPackageFolder() + "/demos_cpp"; }
std::string DemoPythonFolder() { return MainPythonPackageFolder() + "/demos_python"; }
std::string MarkdownDocFolder() { return MainPythonPackageFolder() + "/doc/imgui_bundle_demo_parts"; }


// memoized function
std::string ReadCode(const std::string& filename)
{
    static std::unordered_map<std::string, std::string> gFileContents;
    if (gFileContents.find(filename) != gFileContents.end())
        return gFileContents.at(filename);

    if (std::filesystem::exists(filename))
        gFileContents[filename] = fplus::read_text_file(filename)();
    else
        gFileContents[filename] = "";

    return gFileContents.at(filename);
}


std::string ReadCppCode(const std::string& demo_file_path)
{
    return ReadCode(DemoCppFolder() + "/" + demo_file_path + ".cpp");
}

std::string ReadPythonCode(const std::string& demo_file_path)
{
    return ReadCode(DemoPythonFolder() + "/" + demo_file_path + ".py");
}

std::string ReadMarkdownDoc(const std::string& doc_file_name)
{
    std::string r = ReadCode(MarkdownDocFolder() + "/" + doc_file_name + ".adoc.md");
    return r;
}


void ShowPythonVsCppCode(const std::string& pythonCode, const std::string& cppCode, int nbLines)
{
    ImGui::PushID(pythonCode.c_str());

    const bool flagHalfWidth = !pythonCode.empty() && !cppCode.empty();

    Snippets::SnippetData snippetCpp, snippetPython;

    snippetCpp.Code = cppCode;
    snippetCpp.DisplayedFilename = "C++ code";
    snippetCpp.Language = Snippets::SnippetLanguage::Cpp;
    snippetCpp.HeightInLines = nbLines;
    snippetCpp.MaxHeightInLines = nbLines;

    snippetPython.Code = pythonCode;
    snippetPython.DisplayedFilename = "Python code";
    snippetPython.Language = Snippets::SnippetLanguage::Python;
    snippetPython.HeightInLines = nbLines;
    snippetPython.MaxHeightInLines = nbLines;

    Snippets::ShowSideBySideSnippets(snippetCpp, snippetPython);
    ImGui::PopID();
}


void ShowPythonVsCppFile(const char* demo_file_path, int nbLines)
{
    std::string cpp_code = ReadCppCode(demo_file_path);
    std::string python_code = ReadPythonCode(demo_file_path);
    ShowPythonVsCppCode(python_code.c_str(), cpp_code.c_str(), nbLines);
}

#ifndef __EMSCRIPTEN__
std::string DemoExeFile(const std::string& demoName)
{
#ifndef HELLOIMGUI_INSIDE_APPLE_BUNDLE
        std::string exeFolder = wai_getExecutableFolder_string();
        std::string exeFile = exeFolder + "/" + demoName;
    #ifdef _WIN32
        exeFile += ".exe";
    #endif
        return exeFile;
#else
    std::string exeFolder = wai_getExecutableFolder_string();
    std::string appsBundlesFolder = exeFolder + "/../../../";
    std::string exeFile = appsBundlesFolder + demoName + ".app/Contents/MacOS/" + demoName;
    return exeFile;
#endif
}
#endif

bool HasDemoExeFile(const std::string& demoName)
{
    std::string exeFile = DemoExeFile(demoName);
    return std::filesystem::exists(exeFile);
}


#ifdef __EMSCRIPTEN__
#include <emscripten.h>
bool SpawnDemo(const std::string& demoName)
{
    // This is for emscripten
    std::string jsCommandTemplate = R"(
        window.open("{demoName}.html", "hello", "width=900,height=600");
    )";
    std::string jsCommand = fplus::replace_tokens<std::string>("{demoName}", demoName, jsCommandTemplate);
    printf("%s\n", jsCommand.c_str());
    emscripten_run_script(jsCommand.c_str());
    return true;
}

#else

bool SpawnDemo(const std::string& demoName)
{
    std::string exeFile = DemoExeFile(demoName);

    if (std::filesystem::exists(exeFile))
    {
        const char *command_line[2] = {exeFile.c_str(), NULL};
        struct subprocess_s subprocess;
        #ifdef __linux__
            int system_command_result = system(command_line[0]);
            IM_ASSERT(system_command_result == 0);
        #else
            subprocess_create(command_line, subprocess_option_no_window | subprocess_option_inherit_environment, &subprocess);
        #endif
        return true;
    }
    else
        return false;
}
#endif // #ifdef __EMSCRIPTEN__

void ShowMarkdownDocFile(const std::string& doc_file_name)
{
    auto code = ReadMarkdownDoc(doc_file_name);
    ImGuiMd::RenderUnindented(code);
}


// [sub section]  BrowseToUrl()
// A platform specific utility to open an url in a browser
// (especially useful with emscripten version)
// Specific per platform includes for BrowseToUrl
#if defined(__EMSCRIPTEN__)
#include <emscripten.h>
#elif defined(_WIN32)
#include <windows.h>
#include <Shellapi.h>
#elif defined(__APPLE__)
#include <TargetConditionals.h>
#endif

void BrowseToUrl(const char *url)
{
#if defined(__EMSCRIPTEN__)
    char js_command[1024];
    snprintf(js_command, 1024, "window.open(\"%s\");", url);
    emscripten_run_script(js_command);
#elif defined(_WIN32)
    ShellExecuteA( NULL, "open", url, NULL, NULL, SW_SHOWNORMAL );
#elif TARGET_OS_IPHONE
    // Nothing on iOS
#elif TARGET_OS_OSX
    char cmd[1024];
    snprintf(cmd, 1024, "open %s", url);
    system(cmd);
#elif defined(__linux__)
    char cmd[1024];
    snprintf(cmd, 1024, "xdg-open %s", url);
    int r = system(cmd);
    (void) r;
#endif
}

void BrowseToUrl(const std::string& url)
{
    BrowseToUrl(url.c_str());
}

File: bindings/imgui_bundle/demos_python/demo_utils/api_demos.py
import os
from typing import Callable
from imgui_bundle import immapp, imgui, imgui_md, hello_imgui
from imgui_bundle.demos_python.demo_utils.functional_utils import memoize


GuiFunction = Callable[[], None]


def main_python_package_folder() -> str:
    this_dir = os.path.dirname(__file__)
    this_dir = this_dir.replace("\\", "/")

    items = this_dir.split("/")

    for n in reversed(range(len(items))):
        parent_folder = "/".join(items[:n])
        if os.path.isfile(parent_folder + "/hello_imgui.pyi"):
            return parent_folder

    raise Exception("Cannot find main python package!")


def demos_assets_folder() -> str:
    r = main_python_package_folder() + "/demos_assets"
    return r


def demos_cpp_folder() -> str:
    return main_python_package_folder() + "/demos_cpp"


def demos_python_folder() -> str:
    return main_python_package_folder() + "/demos_python"


def markdown_doc_folder() -> str:
    return main_python_package_folder() + "/doc/imgui_bundle_demo_parts"


def set_hello_imgui_demo_assets_folder():
    hello_imgui.set_assets_folder(demos_assets_folder())


def show_python_vs_cpp_code(python_code: str, cpp_code: str, nb_lines: int = 0):
    imgui.push_id(python_code)

    snippet_cpp: immapp.snippets.SnippetData = immapp.snippets.SnippetData()  # type: ignore
    snippet_cpp.code = cpp_code
    snippet_cpp.displayed_filename = "C++ code"
    snippet_cpp.height_in_lines = nb_lines
    snippet_cpp.max_height_in_lines = nb_lines

    snippet_python: immapp.snippets.SnippetData = immapp.snippets.SnippetData()  # type: ignore
    snippet_python.code = python_code
    snippet_python.displayed_filename = "Python code"
    snippet_python.height_in_lines = nb_lines
    snippet_python.max_height_in_lines = nb_lines

    immapp.snippets.show_side_by_side_snippets(snippet_python, snippet_cpp)

    imgui.pop_id()


def show_python_vs_cpp_and_run(
    python_gui_function, cpp_code: str, nb_lines: int = 0
) -> None:
    import inspect

    python_code = inspect.getsource(python_gui_function)
    show_python_vs_cpp_code(python_code, cpp_code, nb_lines)
    python_gui_function()


def show_python_vs_cpp_file(demo_file_path: str, nb_lines: int = 0) -> None:
    cpp_code = read_cpp_code(demo_file_path)
    python_code = read_python_code(demo_file_path)
    show_python_vs_cpp_code(python_code, cpp_code, nb_lines)


def show_markdown_file(doc_filename: str) -> None:
    code = read_markdown_code(doc_filename)
    imgui_md.render_unindented(code)


@memoize  # type: ignore
def read_code(filename: str) -> str:
    if not os.path.isfile(filename):
        return ""
    with open(filename, encoding="utf8") as f:
        r = f.read()
        return r


def read_cpp_code(demo_file_path: str) -> str:
    file_abs = demos_cpp_folder() + "/" + demo_file_path + ".cpp"
    code: str = read_code(file_abs)  # type: ignore
    return code


def read_python_code(demo_file_path: str) -> str:
    file_abs = demos_python_folder() + "/" + demo_file_path + ".py"
    code: str = read_code(file_abs)  # type: ignore
    return code


def read_markdown_code(doc_filename: str) -> str:
    doc_file = markdown_doc_folder() + "/" + doc_filename + ".adoc.md"
    r: str = read_code(doc_file)  # type: ignore
    return r
